diff --git a/xbmc/cores/playercorefactory/PlayerCoreFactory.h b/xbmc/cores/playercorefactory/PlayerCoreFactory.h
index 9e964ba..4e21b3f 100644
--- a/xbmc/cores/playercorefactory/PlayerCoreFactory.h
+++ b/xbmc/cores/playercorefactory/PlayerCoreFactory.h
@@ -42,12 +42,8 @@ enum EPLAYERCORES
   EPC_DVDPLAYER,
   EPC_MPLAYER,
   EPC_PAPLAYER,
-#if defined(HAS_AMLPLAYER)
   EPC_AMLPLAYER,
-#endif
-#if defined(HAS_OMXPLAYER)
   EPC_OMXPLAYER,
-#endif
   EPC_EXTPLAYER,
   EPC_UPNPPLAYER,
 };
@@ -58,12 +54,8 @@ const PLAYERCOREID PCID_NONE = EPC_NONE;
 const PLAYERCOREID PCID_DVDPLAYER = EPC_DVDPLAYER;
 const PLAYERCOREID PCID_MPLAYER = EPC_MPLAYER;
 const PLAYERCOREID PCID_PAPLAYER = EPC_PAPLAYER;
-#if defined(HAS_AMLPLAYER)
 const PLAYERCOREID PCID_AMLPLAYER = EPC_AMLPLAYER;
-#endif
-#if defined(HAS_OMXPLAYER)
 const PLAYERCOREID PCID_OMXPLAYER = EPC_OMXPLAYER;
-#endif
 
 class CPlayerCoreFactory
 {
